source/agora/network/NetworkManager.d

```diff
@@ -57,31 +57,290 @@ mixin AddLogger!();
 /// Ditto
 public class NetworkManager
 {
+    /// Node information
+    private static struct NodeConnInfo
+    {
+        /// Address of the node
+        Address address;
+
+        /// Is this node a Validator
+        bool is_validator;
+
+        /// Public key of the node, only set if the node is a Validator
+        PublicKey key;
+
+        /// Client
+        NetworkClient client;
+    }
+
+    /***************************************************************************
+
+        Establishes a connection with a Node address in a new task.
+        The BanManager in the nesting class is periodically queried.
+
+    ***************************************************************************/
+
+    private class ConnectionTask
+    {
+        /// Address to connect to
+        private const Address address;
+
+        /// Called when we've connected and determined if this is
+        /// a FullNode / Validator
+        private void delegate (scope ref NodeConnInfo) onHandshakeComplete;
+
+        /// Called when we could not connect, and the address has been banned
+        private void delegate (in Address) onFailedConnection;
+
+
+        /***********************************************************************
+
+            Ctor
+
+            Params:
+                address = the adddress of the node
+                onHandshakeComplete = called when we've successfully connected
+                                      and determined the type of the Node
+
+                onFailedConnection = called when we couldn't connect with the
+                                     node and the node was banned
+
+        ***********************************************************************/
+
+        public this (in Address address,
+            void delegate (scope ref NodeConnInfo node) onHandshakeComplete,
+            void delegate (in Address address) onFailedConnection)
+        {
+            this.address = address;
+            this.onHandshakeComplete = onHandshakeComplete;
+            this.onFailedConnection = onFailedConnection;
+            this.outer.taskman.runTask(&this.connect);
+        }
+
+        /***********************************************************************
+
+            Repeatedly attempt connecting to an address, and try to determine
+            if this is a FullNode or a Validator.
+
+            After each connection failure / request failure the ban manager is
+            queried, and if this address was banned then 'onFailedConnection()'
+            will be called and the task will be killed.
+
+            If we've successfully connected and determined the Node type,
+            'onHandshakeComplete' is called and the task is killed.
+
+        ***********************************************************************/
+
+        private void connect ()
+        {
+            API api;
```

Comment:
Side note: This is an issue that came up when trying to improve the REST module / implementing the TCP protocol.
It gives a pretty bad API to expose disconnections to the client by forcing the user to re-create a client object, so I ended up handling it internally and have auto-reconnect (and potential delegates for failures).

In other words: A client is tied to an address, not a connection.

Reply:
Yeah. I really hate this part of vibe.d.